File: rectangle.cpp
//------------------------------------------------------------------------------
// rectangle.cpp - содержит методы прямоугольника
//------------------------------------------------------------------------------

#include "rectangle.h"

//------------------------------------------------------------------------------
// Ввод параметров прямоугольника из файла
void Rectangle::In(ifstream &ifst) {
    ifst >> x_left >> y_left >> x_right >> y_right;
}

// Случайный ввод параметров прямоугольника
void Rectangle::InRnd() {
    x_left = Shape::rnd20.Get();
    y_left = Shape::rnd20.Get();
    int distance_x = Shape::rnd20.Get();
    int distance_y = Shape::rnd20.Get();
    x_right = x_left + distance_x;
    y_right = y_left - distance_y;
}

//------------------------------------------------------------------------------
// Вывод параметров прямоугольника в форматируемый поток
void Rectangle::Out(ofstream &ofst) {
    ofst << "It is Rectangle: x upper left = " << x_left << ", y upper left = " << y_left
         << ", x lower right = " << x_right << ", y lower right =" << y_right
         << ". Perimeter = " << Perimeter() << "\n";
}

//------------------------------------------------------------------------------
// Вычисление периметра прямоугольника
double Rectangle::Perimeter() {
    return 2.0 * (x_right - x_left + y_left - y_left);
}

File: rectangle.py
import random

from shape import *


# ----------------------------------------------
class Rectangle(Shape):
    def __init__(self):
        self.x_left = 0
        self.y_left = 0
        self.x_right = 0
        self.y_right = 0

    def ReadStrArray(self, strArray, i):
        # должно быт как минимум 4 непрочитанных значения в массиве
        if i >= len(strArray) - 3:
            return 0
        self.x_left = int(strArray[i])
        self.y_left = int(strArray[i + 1])
        self.x_right = int(strArray[i + 2])
        self.y_right = int(strArray[i + 3])
        i += 4
        # print("Rectangle: x = ", self.x, " y = ", self.y)
        return i

    def Print(self):
        print("Rectangle: x upper left = ", self.x_left, " y upper left = ", self.y_left, " x lower right = ",
              self.x_right, " y lower right =", self.y_right, ", Perimeter = ", self.Perimeter())
        pass

    def Write(self, ostream):
        ostream.write("Rectangle: x upper left = {}  y upper left= {} "
                      "x lower right = {} y lower right = {}, Perimeter = {}"
                      .format(self.x_left, self.y_left, self.x_right, self.y_right, self.Perimeter()))
        pass

    def Perimeter(self):
        return 2.0 * (self.x_right - self.x_left + self.y_left - self.y_left)
        pass

    def FillRdn(self):
        self.x_left = random.randint(0, 20)
        self.x_left = random.randint(0, 20)
        distance_x = random.randint(0, 20)
        distance_y = random.randint(0, 20)
        self.x_right = self.x_left + distance_x
        self.y_right = self.y_left - distance_y
        pass
